synapse/handlers/appservice.py

```diff
@@ -176,12 +176,45 @@ def query_3pe(self, kind, protocol, fields):
         defer.returnValue(ret)
 
     @defer.inlineCallbacks
-    def get_3pe_protocols(self):
+    def get_3pe_protocols(self, only_protocol=None):
         services = yield self.store.get_app_services()
         protocols = {}
+
+        # Collect up all the individual protocol responses out of the ASes
         for s in services:
             for p in s.protocols:
-                protocols[p] = yield self.appservice_api.get_3pe_protocol(s, p)
+                if only_protocol is not None and p != only_protocol:
+                    continue
+
+                info = yield self.appservice_api.get_3pe_protocol(s, p)
+
+                # Ignore any result that doesn't contain an "instances" list
+                if "instances" not in info:
+                    continue
+                if not isinstance(info["instances"], list):
+                    continue
```

Comment:
Generally you don't do type checking like this. If you're worried that `info["instances"]` isn't going to be an iterable then you can catch the `TypeError` below where you do `list(info["instances"])`. I'm also surprised this doesn't happen in `appservice_api.get_3pe_protocol`.

Unless this is acceptable behaviour for an app service I'd also log